sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -2267,9 +2267,71 @@ pi_result cuda_piEnqueueKernelLaunch(
   assert(command_queue != nullptr);
   assert(command_queue->get_context() == kernel->get_context());
   assert(kernel != nullptr);
+  assert(global_work_offset != nullptr);
   assert(work_dim > 0);
   assert(work_dim < 4);
 
+  // Set the number of threads per block to the number of threads per warp
+  // by default unless user has provided a better number
+  int threadsPerBlock[3] = {32, 1, 1};
+
+  {
+    size_t maxThreadsPerBlock[3] = {};
+    pi_result retError = cuda_piDeviceGetInfo(
+        command_queue->device_, PI_DEVICE_INFO_MAX_WORK_ITEM_SIZES,
+        sizeof(maxThreadsPerBlock), maxThreadsPerBlock, nullptr);
+    assert(retError == PI_SUCCESS);
+    (void)retError;
+    size_t maxWorkGroupSize = 0;
+    retError = cuda_piDeviceGetInfo(
+        command_queue->device_, PI_DEVICE_INFO_MAX_WORK_GROUP_SIZE,
+        sizeof(maxWorkGroupSize), &maxWorkGroupSize, nullptr);
+    assert(retError == PI_SUCCESS);
+
+    if (local_work_size) {
+      for (size_t i = 0; i < work_dim; i++) {
+        if (local_work_size[i] > maxThreadsPerBlock[i])
+          return PI_INVALID_WORK_ITEM_SIZE;
+        // Checks that local work sizes are a divisor of the global work sizes
+        // which includes that the local work sizes are neither larger than the
+        // global work sizes and not 0.
+        if (0u == local_work_size[i])
+          return PI_INVALID_WORK_GROUP_SIZE;
+        if (0u != (global_work_size[i] % local_work_size[i]))
+          return PI_INVALID_WORK_GROUP_SIZE;
+        threadsPerBlock[i] = static_cast<int>(local_work_size[i]);
+      }
+      if (maxWorkGroupSize < size_t(threadsPerBlock[0] * threadsPerBlock[1] *
+                                    threadsPerBlock[2])) {
+        return PI_INVALID_WORK_GROUP_SIZE;
+      }
+    } else {
+      // Determine local work sizes that result in uniform work groups.
+      for (size_t i = 0; i < work_dim; i++) {
+        threadsPerBlock[i] =
+            std::min(static_cast<int>(maxThreadsPerBlock[i]),
```

Comment:
I realized that if work_dim > 1, then iterations of the loop at L2310 do not make any useful work for i = 1 and i=2 because at the beginning of the function threadsPerBlock is initialized as {32, 1, 1}. Also, global_work_size[i] is always > 0. 
Thus for i = 1 and i = 2:
    tmp1=min(threadsPerBlock[i] /* == 1*/, global_work_size[i] /* >0 */); // tmp is always set to 1
    threadsPerBlock[i]  = min(maxThreadsPerBlock[i], tmp1); // always set to 1.

Reply:
Yes. I wrote this as general code and then left it to keep working should the default values change. This pessimizes performance.

I'll remove the unncessary loop iterations in a way that will make it easy to reintroduce the loop if necessary later.

Reply:
I simplified the code and added a comment about only processing the first dimension due to the default values.